### PR TITLE
log_backup: add lock for truncating

### LINKED ISSUE #49414: log-backup: add mutual execution for log truncating.
## Enhancement
Sometimes, if we are running parallel truncating tasks, we may encounter some strange things: say, a meta file being truncated in one truncate task, but yet another one also wants to delete it. The latter one may finish with a failed deletion and will greatly confuse our users.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #49414 

Problem Summary:
See the issue.

### What changed and how does it work?
This PR added a lock while we are doing truncating. 
The lock will be send with metadata of the node that locks the storage, so other can read the lock and decide whether it is valid.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
image
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Concurrent run of log truncate will cause fail now.
```

### CODE REVIEW

File: br/pkg/storage/locking.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
+package storage
```
Comment (BornChanger) [inline br/pkg/storage/locking.go:1]: license claim missing

File: br/pkg/storage/locking_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package storage_test
```
Comment (BornChanger) [inline br/pkg/storage/locking_test.go:1]: license claim missing

File: br/pkg/storage/locking.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// MakeLockMeta creates a LockMeta by the current node's metadata.
+// Including current time and hostname, etc..
+func MakeLockMeta(hint string) LockMeta {
+	hname, err := os.Hostname()
+	if err != nil {
+		hname = fmt.Sprintf("UnknownHost(err=%s)", err)
+	}
+	now := time.Now()
```
Comment (Leavrth) [inline br/pkg/storage/locking.go:44]: does it need to get ts from PD？

**Review by Leavrth — APPROVED** [review]

File: br/pkg/storage/locking.go (hunk 1 above)
Comment (BornChanger) [reply br/pkg/storage/locking.go:1]: please also consolidate the br locking logic here?

File: br/pkg/storage/locking.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+			return err
+		}
+		return ErrLocked{Meta: meta}
+	}
+
+	meta := MakeLockMeta(hint)
+	if err := putLockMeta(ctx, storage, path, meta); err != nil {
+		return err
+	}
+	return nil
```
Comment (BornChanger) [inline br/pkg/storage/locking.go:107]:
```suggestion
	return putLockMeta(ctx, storage, path, meta)
```

**Review by BornChanger — APPROVED** [review]

**Pull request merged**